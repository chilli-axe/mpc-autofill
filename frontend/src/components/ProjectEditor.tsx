/**
 * This component represents the complete MPC Autofill project editor, ready to
 * drop into a page (as the only component). Must be wrapped with a Redux provider.
 */

import React, { useEffect } from "react";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { RibbonHeight } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { NoBackendDefault } from "@/components/NoBackendDefault";
import { OverflowCol } from "@/components/OverflowCol";
import { Ribbon } from "@/components/Ribbon";
import { SelectedImagesRibbon } from "@/features/bulkManagement/SelectedImagesRibbon";
import { CardGrid } from "@/features/card/CardGrid";
import { CommonCardback } from "@/features/card/CommonCardback";
import { Export } from "@/features/export/Export";
import { FinishedMyProject } from "@/features/export/FinishedMyProjectModal";
import { FinishSettings } from "@/features/finishSettings/FinishSettings";
import { Import } from "@/features/import/Import";
import { SearchSettings } from "@/features/searchSettings/SearchSettings";
import { Status } from "@/features/status/Status";
import { useBackendConfigured } from "@/store/slices/backendSlice";
import {
  selectIsProjectEmpty,
  selectProjectCardback,
} from "@/store/slices/projectSlice";

function ProjectEditor() {
  // TODO: should we periodically ping the backend to make sure it's still alive?
  //# region queries and hooks

  const backendConfigured = useBackendConfigured();
  const cardback = useAppSelector(selectProjectCardback);
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);

  //# endregion

  //# region effects

  /**
   * Ask the user for confirmation before they close the page if their project has any cards in it.
   */
  useEffect(() => {
    const handler = (event: BeforeUnloadEvent) => {
      if (!isProjectEmpty) {
        event.preventDefault();
        return false;
      }
    };
    window.addEventListener("beforeunload", handler);
    return () => {
      window.removeEventListener("beforeunload", handler);
    };
  }, [isProjectEmpty]);

  //# endregion

  return (
    <>
      {backendConfigured ? (
        <Row className="g-0">
          <Ribbon className="g-0">
            <SelectedImagesRibbon />
          </Ribbon>
          <OverflowCol
            lg={8}
            md={8}
            sm={6}
            xs={6}
            data-testid="left-panel"
            heightDelta={RibbonHeight}
          >
            <CardGrid />
          </OverflowCol>
          <OverflowCol
            data-testid="right-panel"
            lg={4}
            md={4}
            sm={6}
            xs={6}
            style={{ zIndex: 1 }}
            className="px-2"
            heightDelta={RibbonHeight}
          >
            <Status />
            <Row className="g-0 pt-2">
              <FinishSettings />
            </Row>
            <Row className="g-0 pt-2">
              <SearchSettings />
            </Row>
            <Row className="g-0 pt-2">
              <Col lg={7} md={12} sm={12} xs={12}>
                <Import />
              </Col>
              <Col lg={5} md={12} sm={12} xs={12}>
                <Export />
              </Col>
              <FinishedMyProject />
            </Row>
            <Col className="g-0 pt-2" lg={{ span: 8, offset: 2 }} md={12}>
              <CommonCardback selectedImage={cardback} />
            </Col>
          </OverflowCol>
        </Row>
      ) : (
        <NoBackendDefault />
      )}
    </>
  );
}

export default ProjectEditor;
