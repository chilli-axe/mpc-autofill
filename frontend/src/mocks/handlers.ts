import { http, HttpResponse } from "msw";

import { Card, Cardback, Token } from "@/common/constants";
import { Campaign, Supporter, SupporterTier } from "@/common/schema_types";
import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  cardDocument4,
  cardDocument5,
  cardDocument6,
  localBackend,
  sourceDocument1,
  sourceDocument2,
  sourceDocument3,
} from "@/common/test-constants";

const createError = (name: string) => ({
  name,
  message: "A message that describes the error",
});

/**
 * Not including the correct leading and trailing slashes can break things.
 * This little helper function ensures the given relative API route is associated
 * with the local backend URL correctly.
 * TODO: not sure how true the above statement is as of MSW 2.7
 */
function buildRoute(route: string) {
  const re = /^\/?(.*?)\/?$/g;
  return `${localBackend.url}/${(re.exec(route) ?? ["", ""])[1]}`;
}

//# region source

export const sourceDocumentsNoResults = http.get(buildRoute("2/sources/"), () =>
  HttpResponse.json({ results: {} }, { status: 200 })
);

export const sourceDocumentsOneResult = http.get(buildRoute("2/sources/"), () =>
  HttpResponse.json(
    {
      results: {
        [sourceDocument1.pk]: sourceDocument1,
      },
    },
    { status: 200 }
  )
);

export const sourceDocumentsThreeResults = http.get(
  buildRoute("2/sources/"),
  () =>
    HttpResponse.json(
      {
        results: {
          [sourceDocument1.pk]: sourceDocument1,
          [sourceDocument2.pk]: sourceDocument2,
          [sourceDocument3.pk]: sourceDocument3,
        },
      },
      { status: 200 }
    )
);

export const sourceDocumentsServerError = http.get(
  buildRoute("2/sources/"),
  () => HttpResponse.json(createError("2/sources"), { status: 500 })
);

//# endregion

//# region card

export const cardDocumentsNoResults = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json({ results: {} }, { status: 200 })
);

export const cardDocumentsOneResult = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json(
    {
      results: {
        [cardDocument1.identifier]: cardDocument1,
      },
    },
    { status: 200 }
  )
);

export const cardDocumentsThreeResults = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json(
    {
      results: {
        [cardDocument1.identifier]: cardDocument1,
        [cardDocument2.identifier]: cardDocument2,
        [cardDocument3.identifier]: cardDocument3,
      },
    },
    { status: 200 }
  )
);

export const cardDocumentsFourResults = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json(
    {
      results: {
        [cardDocument1.identifier]: cardDocument1,
        [cardDocument2.identifier]: cardDocument2,
        [cardDocument3.identifier]: cardDocument3,
        [cardDocument4.identifier]: cardDocument4,
      },
    },
    { status: 200 }
  )
);

export const cardDocumentsSixResults = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json(
    {
      results: {
        [cardDocument1.identifier]: cardDocument1,
        [cardDocument2.identifier]: cardDocument2,
        [cardDocument3.identifier]: cardDocument3,
        [cardDocument4.identifier]: cardDocument4,
        [cardDocument5.identifier]: cardDocument5,
        [cardDocument6.identifier]: cardDocument6,
      },
    },
    { status: 200 }
  )
);

export const cardDocumentsServerError = http.post(buildRoute("2/cards/"), () =>
  HttpResponse.json(createError("2/cards"), { status: 500 })
);

//# endregion

//# region cardback

export const cardbacksNoResults = http.post(buildRoute("2/cardbacks"), () =>
  HttpResponse.json({ cardbacks: [] }, { status: 200 })
);

export const cardbacksOneResult = http.post(buildRoute("2/cardbacks"), () =>
  HttpResponse.json(
    {
      cardbacks: [cardDocument1.identifier],
    },
    { status: 200 }
  )
);

export const cardbacksOneOtherResult = http.post(
  buildRoute("2/cardbacks"),
  () =>
    HttpResponse.json(
      {
        cardbacks: [cardDocument5.identifier],
      },
      { status: 200 }
    )
);

export const cardbacksTwoResults = http.post(buildRoute("2/cardbacks"), () =>
  HttpResponse.json(
    {
      cardbacks: [cardDocument1.identifier, cardDocument2.identifier],
    },
    { status: 200 }
  )
);

export const cardbacksTwoOtherResults = http.post(
  buildRoute("2/cardbacks"),
  () =>
    HttpResponse.json(
      {
        cardbacks: [cardDocument2.identifier, cardDocument3.identifier],
      },
      { status: 200 }
    )
);

export const cardbacksServerError = http.post(buildRoute("2/cardbacks/"), () =>
  HttpResponse.json(createError("2/cardbacks"), { status: 500 })
);

//# endregion

//# region search results

export const searchResultsNoResults = http.post(
  buildRoute("2/editorSearch/"),
  () => HttpResponse.json({ results: {} }, { status: 200 })
);

export const searchResultsOneResult = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          "my search query": {
            CARD: [cardDocument1.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsOneResultCorrectSearchq = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          [cardDocument1.searchq]: {
            CARD: [cardDocument1.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsThreeResults = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          "my search query": {
            CARD: [
              cardDocument1.identifier,
              cardDocument2.identifier,
              cardDocument3.identifier,
            ],
            CARDBACK: [],
            TOKEN: [],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsFourResults = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          "my search query": {
            CARD: [
              cardDocument1.identifier,
              cardDocument2.identifier,
              cardDocument3.identifier,
              cardDocument4.identifier,
            ],
            CARDBACK: [cardDocument5.identifier],
            TOKEN: [cardDocument6.identifier],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsSixResults = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          "query 1": {
            CARD: [cardDocument1.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "query 2": {
            CARD: [cardDocument2.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "query 3": {
            CARD: [cardDocument3.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "query 4": {
            CARD: [cardDocument4.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "query 5": {
            CARD: [],
            CARDBACK: [cardDocument5.identifier],
            TOKEN: [],
          },
          "query 6": {
            CARD: [],
            CARDBACK: [],
            TOKEN: [cardDocument6.identifier],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsForDFCMatchedCards1And4 = http.post(
  buildRoute("2/editorSearch/"),
  () =>
    HttpResponse.json(
      {
        results: {
          "my search query": {
            CARD: [cardDocument1.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "card 3": {
            CARD: [cardDocument3.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
          "card 4": {
            CARD: [cardDocument4.identifier],
            CARDBACK: [],
            TOKEN: [],
          },
        },
      },
      { status: 200 }
    )
);

export const searchResultsServerError = http.post(
  buildRoute("2/editorSearch/"),
  () => HttpResponse.json(createError("2/editorSearch"), { status: 200 })
);

//# endregion

//# region dfc pairs

export const dfcPairsNoResults = http.get(buildRoute("2/DFCPairs/"), () =>
  HttpResponse.json({ dfcPairs: {} }, { status: 200 })
);

export const dfcPairsMatchingCards1And4 = http.get(
  buildRoute("2/DFCPairs/"),
  () =>
    HttpResponse.json(
      { dfcPairs: { ["my search query"]: cardDocument4.name } },
      { status: 200 }
    )
);

export const dfcPairsServerError = http.get(buildRoute("2/DFCPairs/"), () =>
  HttpResponse.json(createError("2/DFCPairs"), { status: 500 })
);

//# endregion

//# region languages

export const languagesNoResults = http.get(buildRoute("2/languages/"), () =>
  HttpResponse.json({ languages: [] }, { status: 200 })
);

export const languagesTwoResults = http.get(buildRoute("2/languages/"), () =>
  HttpResponse.json(
    {
      languages: [
        { name: "English", code: "EN" },
        { name: "French", code: "FR" },
      ],
    },
    { status: 200 }
  )
);

//# endregion

//# region tags

export const tagsNoResults = http.get(buildRoute("2/tags/"), () =>
  HttpResponse.json({ tags: [] }, { status: 200 })
);

export const tagsTwoResults = http.get(buildRoute("2/tags/"), () =>
  HttpResponse.json({ tags: ["Tag 1", "Tag 2"] }, { status: 200 })
);

//# endregion

//# region sample cards

export const sampleCards = http.get(buildRoute("2/sampleCards"), () =>
  HttpResponse.json(
    {
      cards: {
        [Card]: [cardDocument1, cardDocument2, cardDocument3, cardDocument4],
        [Cardback]: [cardDocument5],
        [Token]: [cardDocument6],
      },
    },
    { status: 200 }
  )
);

export const sampleCardsServerError = http.get(
  buildRoute("2/sampleCards/"),
  () => HttpResponse.json(createError("2/sampleCards"), { status: 500 })
);

//# endregion

//# region import sites

export const importSitesNoResults = http.get(buildRoute("2/importSites"), () =>
  HttpResponse.json({ importSites: [] }, { status: 200 })
);

export const importSitesOneResult = http.get(buildRoute("2/importSites"), () =>
  HttpResponse.json(
    { importSites: [{ name: "test", url: "test.com" }] },
    { status: 200 }
  )
);

export const importSitesServerError = http.get(
  buildRoute("2/importSites/"),
  () => HttpResponse.json(createError("2/importSites"), { status: 500 })
);

//# endregion

//# region what's new

export const newCardsFirstPageWithTwoSources = http.get(
  buildRoute("2/newCardsFirstPages"),
  () =>
    HttpResponse.json(
      {
        results: {
          [sourceDocument1.key]: {
            source: sourceDocument1,
            hits: 4,
            pages: 2,
            cards: [cardDocument1, cardDocument2],
          },
          [sourceDocument2.key]: {
            source: sourceDocument2,
            hits: 1,
            pages: 1,
            cards: [cardDocument5],
          },
        },
      },
      { status: 200 }
    )
);

export const newCardsFirstPageNoResults = http.get(
  buildRoute("2/newCardsFirstPages"),
  () =>
    HttpResponse.json(
      {
        results: {},
      },
      { status: 200 }
    )
);

export const newCardsPageForSource1 = http.get(
  buildRoute(`2/newCardsPage?source=${sourceDocument1.key}&page=2`),
  () =>
    HttpResponse.json(
      { cards: [cardDocument3, cardDocument4] },
      { status: 200 }
    )
);

export const newCardsFirstPageServerError = http.get(
  buildRoute("2/newCardsFirstPages"),
  () => HttpResponse.json(createError("2/newCardsFirstPage"), { status: 500 })
);

//# endregion

//# region backend info

export const backendInfo = http.get(buildRoute("2/info"), () =>
  HttpResponse.json(
    {
      info: {
        name: "Test Site",
        description: "Test runner site",
        email: "test@test.com",
        reddit: "reddit.com",
        discord: "discord.com",
      },
    },
    { status: 200 }
  )
);
export const patreon = http.get(buildRoute("2/patreon"), () =>
  HttpResponse.json(
    {
      patreon: {
        campaign: null,
        members: [],
        tiers: null,
        url: null,
      },
    },
    { status: 200 }
  )
);

export const backendInfoServerError = http.get(buildRoute("2/info/"), () =>
  HttpResponse.json(createError("2/info"), { status: 500 })
);

//# endregion

//# region health

export const searchEngineHealthy = http.get(
  buildRoute("2/searchEngineHealth/"),
  () => HttpResponse.json({ online: true }, { status: 200 })
);

//# endregion

//# region presets

export const defaultHandlers = [
  sourceDocumentsNoResults,
  cardDocumentsNoResults,
  cardbacksNoResults,
  searchResultsNoResults,
  dfcPairsNoResults,
  languagesTwoResults,
  tagsNoResults,
  importSitesOneResult,
  sampleCards,
  backendInfo,
  patreon,
  searchEngineHealthy,
];

//# endregion
