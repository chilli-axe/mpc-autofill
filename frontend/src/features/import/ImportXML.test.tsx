import { screen } from "@testing-library/react";

import {
  Back,
  Card,
  Front,
  S30,
  SelectedImageSeparator,
} from "@/common/constants";
import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
  cardDocument4,
  cardDocument5,
  cardDocument6,
} from "@/common/test-constants";
import {
  expectCardbackSlotState,
  expectCardGridSlotState,
  expectCardSlotToExist,
  importText,
  importXML,
  openImportXMLModal,
  renderWithProviders,
} from "@/common/test-utils";
import ProjectEditor from "@/components/ProjectEditor";
import {
  cardbacksTwoOtherResults,
  cardDocumentsSixResults,
  cardDocumentsThreeResults,
  defaultHandlers,
  searchResultsFourResults,
  searchResultsOneResult,
  searchResultsSixResults,
  searchResultsThreeResults,
  sourceDocumentsOneResult,
  sourceDocumentsThreeResults,
} from "@/mocks/handlers";
import { server } from "@/mocks/server";

const preloadedState = {
  project: {
    members: [],
    cardback: cardDocument2.identifier,
    mostRecentlySelectedSlot: null,
  },
};

//# region snapshot tests

test("the html structure of XML importer", async () => {
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await openImportXMLModal();

  expect(screen.getByTestId("import-xml")).toMatchSnapshot();
});

//# endregion

test("importing one card by XML into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  await importXML(
    `<order>
      <details>
        <quantity>1</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>0</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <cardback>${cardDocument3.identifier}</cardback>
    </order>`
  );

  // a card slot should have been created and it should retain the xml's cardback, not the project cardback
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument3.name, 2, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing multiple instances of one card by XML into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import two instances of a card
  await importXML(
    `<order>
      <details>
        <quantity>2</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <cardback>${cardDocument2.identifier}</cardback>
    </order>`
  );

  // two card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing one specific card version by XML into an empty project", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsThreeResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  await importXML(
    `<order>
      <details>
        <quantity>1</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument3.identifier}</id>
          <slots>0</slots>
          <name>${cardDocument3.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <cardback>${cardDocument2.identifier}</cardback>
    </order>`
  );

  // a card slot should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 3);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing one card of each type into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsThreeResults,
    searchResultsSixResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import one card of each type
  await importXML(
    `<order>
      <details>
        <quantity>3</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>0</slots>
          <name>${cardDocument1.name}</name>
          <query>query 1</query>
        </card>
        <card>
          <id>${cardDocument6.identifier}</id>
          <slots>1</slots>
          <name>${cardDocument6.name}</name>
          <query>t:query 6</query>
        </card>
        <card>
          <id>${cardDocument5.identifier}</id>
          <slots>2</slots>
          <name>${cardDocument5.name}</name>
          <query>b:query 5</query>
        </card>
      </fronts>
      <cardback>${cardDocument3.identifier}</cardback>
    </order>`
  );

  // three card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument3.name, 2, 2);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument6.name, 1, 1);
  await expectCardGridSlotState(2, Back, cardDocument3.name, 2, 2);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument5.name, 1, 1);
  await expectCardGridSlotState(3, Back, cardDocument3.name, 2, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing a more complex XML into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsFourResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a few cards
  await importXML(
    `<order>
      <details>
        <quantity>3</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument3.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument3.name}</name>
          <query>my search query</query>
        </card>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>2</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <backs>
        <card>
          <id>${cardDocument4.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument4.name}</name>
          <query>my search query</query>
        </card>
      </backs>
      <cardback>${cardDocument2.identifier}</cardback>
    </order>`
  );

  // three card slots should have been created
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(1, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(2, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument1.name, 1, 4);
  await expectCardGridSlotState(3, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing an XML with gaps into an empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsFourResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a few cards
  await importXML(
    `<order>
      <details>
        <quantity>4</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument3.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument3.name}</name>
          <query>my search query</query>
        </card>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>3</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <backs>
        <card>
          <id>${cardDocument4.identifier}</id>
          <slots>0,3</slots>
          <name>${cardDocument4.name}</name>
          <query>my search query</query>
        </card>
      </backs>
      <cardback>${cardDocument2.identifier}</cardback>
    </order>`
  );

  // four card slots should have been created, but only three will have images selected
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(1, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(2, Back, cardDocument2.name, 1, 2);
  await expectCardSlotToExist(4);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(4, Front, cardDocument1.name, 1, 4);
  await expectCardGridSlotState(4, Back, cardDocument4.name, 4, 4);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("importing a more complex XML into a non-empty project", async () => {
  server.use(
    cardDocumentsSixResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsFourResults,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, {
    preloadedState: {
      ...preloadedState,
      project: {
        members: [],
        cardback: cardDocument2.identifier,
        mostRecentlySelectedSlot: null,
      },
    },
  });

  // this used to preload the redux state, but with the shift to listeners,
  // we have to add the first card manually like this.
  await importText(
    `1x my search query${SelectedImageSeparator}${cardDocument1.identifier}`
  );
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 4);
  await expectCardGridSlotState(1, Back, cardDocument2.name, 1, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a few more cards
  await importXML(
    `<order>
      <details>
        <quantity>3</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument3.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument3.name}</name>
          <query>my search query</query>
        </card>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>2</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <backs>
        <card>
          <id>${cardDocument4.identifier}</id>
          <slots>0,1</slots>
          <name>${cardDocument4.name}</name>
          <query>my search query</query>
        </card>
      </backs>
      <cardback>${cardDocument3.identifier}</cardback>
    </order>`
  );

  // three card slots should have been created
  await expectCardSlotToExist(2);
  await expectCardGridSlotState(2, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(2, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(3);
  await expectCardGridSlotState(3, Front, cardDocument3.name, 3, 4);
  await expectCardGridSlotState(3, Back, cardDocument4.name, 4, 4);
  await expectCardSlotToExist(4);
  await expectCardGridSlotState(4, Front, cardDocument1.name, 1, 4);
  await expectCardGridSlotState(4, Back, cardDocument3.name, 2, 2);
  await expectCardbackSlotState(cardDocument2.name, 1, 2); // should not have changed
});

test("import an XML and retain its cardback", async () => {
  server.use(
    cardDocumentsThreeResults,
    cardbacksTwoOtherResults,
    sourceDocumentsOneResult,
    searchResultsOneResult,
    ...defaultHandlers
  );
  renderWithProviders(<ProjectEditor />, { preloadedState });

  await expectCardbackSlotState(cardDocument2.name, 1, 2);

  // import a card
  await importXML(
    `<order>
      <details>
        <quantity>1</quantity>
        <bracket>18</bracket>
        <stock>${S30}</stock>
        <foil>false</foil>
      </details>
      <fronts>
        <card>
          <id>${cardDocument1.identifier}</id>
          <slots>0</slots>
          <name>${cardDocument1.name}</name>
          <query>my search query</query>
        </card>
      </fronts>
      <cardback>${cardDocument3.identifier}</cardback>
    </order>`,
    true
  );

  // a card slot should have been created and it should retain the xml's cardback, not the project cardback
  await expectCardSlotToExist(1);
  await expectCardGridSlotState(1, Front, cardDocument1.name, 1, 1);
  await expectCardGridSlotState(1, Back, cardDocument3.name, 2, 2);
  // the project cardback should also have been updated
  await expectCardbackSlotState(cardDocument3.name, 2, 2);
});
