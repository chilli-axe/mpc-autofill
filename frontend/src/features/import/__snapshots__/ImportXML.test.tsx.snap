// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of XML importer 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 20px;
  border-width: 2px;
  border-radius: 2px;
  border-color: #eeeeee;
  border-style: dashed;
  color: #bdbdbd;
  outline: none;
  -webkit-transition: border 0.24s ease-in-out;
  transition: border 0.24s ease-in-out;
  cursor: pointer;
}

<div
  class="modal-dialog modal-dialog-scrollable"
  data-testid="import-xml"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Add Cards â€” XML
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <p>
        Upload an XML file of cards to add to the project.
      </p>
      <p>
        The 
        MPC Autofill
         website can generate an XML file representing your project, and the 
        MPC Autofill
         desktop tool which auto-fills your order into
         
        <a
          href="https://www.makeplayingcards.com"
          target="_blank"
        >
          MakePlayingCards.com
        </a>
         
        expects a file in this format.
      </p>
      <div
        class="btn toggle btn-md off btn-info"
        role="button"
        style="width: 100%; height: 38px;"
      >
        <div
          class="toggle-group"
        >
          <span
            class="btn toggle-on btn-md flex-centre btn-success"
          >
            Use XML Cardback
          </span>
          <span
            class="btn toggle-off btn-md flex-centre btn-info"
          >
            Retain Selected Cardback
          </span>
          <span
            class="toggle-handle btn btn-md btn-default"
          />
        </div>
      </div>
      <div
        class="pt-3"
      >
        <div
          class="btn toggle btn-md off btn-info"
          role="button"
          style="width: 100%; height: 38px;"
        >
          <div
            class="toggle-group"
          >
            <span
              class="btn toggle-on btn-md flex-centre btn-success"
            >
              Use XML Finish Settings
            </span>
            <span
              class="btn toggle-off btn-md flex-centre btn-info"
            >
              Retain Selected Finish Settings
            </span>
            <span
              class="toggle-handle btn btn-md btn-default"
            />
          </div>
        </div>
      </div>
      <hr />
      <div
        class="container"
      >
        <div
          aria-label="import-xml"
          class="c0"
          role="presentation"
          tabindex="0"
        >
          <input
            accept="text/xml,.xml"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          Drag and drop a file here, or click to select a file.
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
